keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2333 @@
+PodSpec which would be enforced by admission checks in the api-server and
+kubelet. In other words, this would be purely informational, aimed for the
+container runtime.
+
+```diff
++ // Kubernetes-managed QoS resources. These are only informational to the runtime
++ // and no CRI requests should fail because of these.
++const (
++       // QOSResourcePodQOS is the Kubernetes Pod Quality of Service Class.
++       // Possible values are "Guaranteed", "Burstable" and "BestEffort".
++       QOSResourcePodQOS = "pod-qos"
++)
+```
+
+##### Kubernetes-managed QoS-class resources
+
+It would be possible to have QoS-class resources that would be managed by
+Kubernetes/kubelet instead of the container runtime. If we specify and manage
+official well-known QoS-class resource names in the API it would be possible to
+specify Kubernetes-internal names that the container runtime would know to
+ignore (or not try to manage itself).
+
+One possible usage-scenario would be pod-level cgroup controls, e.g. cgroup v2
+memory knobs in linux (see
+[KEP-2570: Support Memory QoS with cgroups v2][kep-2570].
+
+Another possible usage could be to allow the user to explicitly specify the
+desired Pod QoS class of the application (instead of implicitly deriving it
+from the resource requests/limits).
+
+##### Container-level memory QoS
+
+Container runtimes could implement an admin-configurable support for memory
+QoS, as an alternative to
+[KEP-2570: Support Memory QoS with cgroups v2][kep-2570].
+
+It would be easy to enable also for example `memory.swap.*` control knobs
+available in Linux cgroups v2 to control swap-usage on a per-container basis.
+
+##### Runtime classes
+
+The QoS-class resources mechanism could be used to replace (or
+re-implement) the current runtime classes.
+
+One benefit of using a QoS-class resource to represent runtime classes would be
+the "automatic" visibility to what runtime classes are available on each node.
+E.g. adding a new runtime class on a node by re-configuration of the runtime
+would be immediately reflected in the node status without any additional admin
+tasks (like node labeling or manual creation of RuntimeClass objects).
+
+##### Splitting Pod QoS Class
+
+Currently the Pod QoS class is an implicit property of a Pod, tied to how the
+resource requests and limits of its containers are specified. QoS-class
+resources could make it possible for users to explicitly specify the Pod QoS
+class in the Pod's `.spec`, making it possible e.g. to define a pod with a
+container whose memory limit is higher than the requests, but that is treated
+by the node as `Guaranteed`.
+
+Taking this idea further, QoS-class resources could also make it possible to
+split several properties implicit in the Pod QoS class (for example: eviction
+behavior or OOM scoring) into separate properties. For example, in the case of
+OOM scoring make it possible for the user to specify a burstable pod with low
+OOM priority (low chance of being killed).
```

Comment:
Can you elaborate how this can be achieved? I am struggling to wrap my head around this.

Reply:
Again, this is speculation for possible future usages, not in any means a fully proven design.

But, for example, the user would specify:

```yaml
spec:
  qosResources:
  - name: k8s.io/podqos
    class: Guaranteed
```

and then the pod would be put as a guaranteed pod in the cgroup hierarchy etc.